### DISCUSSION TITLE
Custom trpc adapter for Chrome Extension with React.js

### DISCUSSION THREAD

**Discussion by harel-tussi** [trpc/trpc#5798] [Ideas]
Hi,
I’m developing a Chrome extension using Manifest V3, with three main components: a React.js-based content script and popup, and a background service worker acting as an API. Currently, the content and popup communicate with the background via chrome.runtime.sendMessage and react-query.

I’d like to integrate tRPC to streamline communication between these components. Is it possible to create a custom tRPC adapter for this setup? Any insights or examples on how to achieve this would be helpful. Thanks!

**Comment by H-Hp** [comment]
Yes, it is possible to integrate tRPC in your Chrome extension setup with a React.js-based content script, popup, and a background service worker. tRPC supports creating custom adapters for different environments, and you can create a custom adapter for the Chrome extension environment.

Here's a high-level overview of how you can achieve this:

1. **Set up tRPC in the Background Service Worker**:
   - Install `@trpc/server` and any other required packages in your background service worker.
   - Define your tRPC router and procedures in the background service worker.
   - Create a custom tRPC adapter that listens for messages from the content script and popup, handles the tRPC requests, and sends back the responses.

2. **Set up tRPC in the Content Script and Popup**:
   - Install `@trpc/client` and any other required packages in your content script and popup projects.
   - Create a custom tRPC adapter that sends messages to the background service worker and receives responses.
   - Create an instance of the tRPC client using the custom adapter.
   - Use the tRPC client in your React components to make tRPC procedure calls.

Here's an example of how you could implement a custom tRPC adapter for the Chrome extension environment:

**Background Service Worker**:

```javascript
// background.js
import { createTRPCProxyAdapter } from '@trpc/server/adapters/proxy';
import { createHTTPServer } from '@trpc/server/http';
import { appRouter } from './trpc-router';

const trpcAdapter = createTRPCProxyAdapter({
  router: appRouter,
  createContext: () => ({}),
});

chrome.runtime.onConnect.addListener((port) => {
  const httpServer = createHTTPServer({
    router: appRouter,
    createContext: () => ({}),
  });

  port.onMessage.addListener((message) => {
    httpServer.handleHTTP(message).then((response) => {
      port.postMessage(response);
    });
  });
});
```

**Content Script / Popup**:

```javascript
// content-script.js or popup.js
import { createTRPCProxyClient } from '@trpc/client/adapters/proxy';
import { appRouter } from './trpc-router'; // Same as in the background service worker

// Create a custom adapter for the Chrome extension environment
const chromeExtensionAdapter = createTRPCProxyClient({
  send: (message) => {
    return new Promise((resolve, reject) => {
      const port = chrome.runtime.connect();
      port.onMessage.addListener((response) => resolve(response));
      port.onDisconnect.addListener(reject);
      port.postMessage(message);
    });
  },
});

// Create an instance of the tRPC client using the custom adapter
const trpc = createTRPCProxyClient({
  adapter: chromeExtensionAdapter,
  router: appRouter,
});

// Use the tRPC client in your React components
function MyComponent() {
  const { data, isLoading } = trpc.useQuery(['someQuery', { input: 'value' }]);

  // ...
}
```

In this example, we're using the `@trpc/server/adapters/proxy` and `@trpc/client/adapters/proxy` adapters provided by tRPC. These adapters allow you to create custom communication channels for your specific environment (in this case, the Chrome extension environment).

On the background service worker side, we're setting up a listener for `chrome.runtime.onConnect` and creating an HTTP server using `createHTTPServer`. When a message is received from the content script or popup, we handle the HTTP request and send back the response.

On the content script/popup side, we're creating a custom adapter that sends messages to the background service worker using `chrome.runtime.connect` and listens for responses. We then create an instance of the tRPC client using this custom adapter and the same router as in the background service worker.

With this setup, you can use tRPC's type-safe APIs and automatic code generation in your React components to communicate with the background service worker seamlessly.

Note that this is a simplified example, and you may need to adjust it based on your specific requirements, such as handling errors, serialization/deserialization, and authentication/authorization. Additionally, you might want to explore other tRPC adapters or create your own custom adapter if the provided ones don't fit your needs.

**Reply by harel-tussi** [reply]
What is createTRPCProxyAdapter? I'm not finding any docs about it

**Reply by spookyuser** [reply]
I feel like this answer was generated by gpt, createTRPCProxyAdapter is depreciated and doesn't seem to have ever had a send parameter. 

You could try this though: https://github.com/janek26/trpc-browser

**Reply by jedwards1211** [reply]
@KATT what do you think about an official documentation page for how to create custom tRPC adapters?  I find myself wanting to use tRPC for all kinds of things...communication with AWS tasks via Redis pubsub, calls to ServiceWorkers, etc.

**Comment by harel-tussi** [comment]
Thank you very much 🙏 
Will try it 🚀

**Comment by wszgrcy** [comment]
https://github.com/wszgrcy/chrome-trpc/blob/master/packages/trpc/readme.md
trpc v11 support
